src/core/mainloop/mainloop.c

```diff
@@ -2003,6 +2005,55 @@ add_entropy_callback(time_t now, const or_options_t *options)
   return ENTROPY_INTERVAL;
 }
 
+/** Periodic callback: if there has been no network usage in a while,
+ * enter a dormant state. */
+static int
+check_network_participation_callback(time_t now, const or_options_t *options)
+{
+  /* If we're a server, we can't become dormant. */
+  if (server_mode(options)) {
+    note_user_activity(now);
+    goto end;
+  }
+
+  /* If we're running an onion service, we can't become dormant. */
+  /* XXXX this would be nice to change, so that we can be dormant with a
+   * service. */
+  if (hs_service_get_num_services() || rend_num_services()) {
+    note_user_activity(now);
+    goto end;
+  }
+
+  /* XXXX Add an option to never become dormant. */
+
+  /* If we have any currently open entry streams other than "linked"
+   * connections used for directory requests, those count as user activity.
+   */
+  /* XXXX make this configurable? */
+  if (connection_get_by_type_nonlinked(CONN_TYPE_AP) != NULL) {
+    note_user_activity(now);
+    goto end;
+  }
+
+  /** Become dormant if there has been no user activity in this long. */
+  /* XXXX make this configurable! */
+#define BECOME_DORMANT_AFTER_INACTIVITY (24*60*60)
+  if (get_last_user_activity_time() + BECOME_DORMANT_AFTER_INACTIVITY >= now) {
+    log_notice(LD_GENERAL, "No user activity in a long time: becoming"
+               " dormant.");
+    set_network_participation(false);
+    rescan_periodic_events(options);
```

Comment:
Shouldn't we instead schedule the event added in an earlier commit or we want this asap?

Reply:
We need this asap, so that we don't run any events that cause us to decide we're non-dormant.  (That shouldn't happen, but still.)

Reply:
> I have a general question.
> 
> Lets assume tor is dormant, and then I change my torrc to become a relay (ORPort let say), the roles should be reloaded to enable bunch of things except that `is_participating_network()` will be false so how do we get out of dormant then without a SIGACTIVE signal?

Oh, that's an important one.  Good catch!  Fix in 9eee3c02d6d568a7b55531e1ad112aba8aa465e0